Tests/PerformanceTest/XCTestCase+InstructionsCount.swift

```diff
@@ -0,0 +1,72 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import _InstructionCounter
+
+extension XCTestCase {
+  private var baselineURL: URL {
+    if let baselineFile = ProcessInfo.processInfo.environment["BASELINE_FILE"] {
+      return URL(fileURLWithPath: baselineFile)
+    } else {
+      return URL(fileURLWithPath: #file)
+        .deletingLastPathComponent()
+        .appendingPathComponent("baselines.json")
+    }
+  }
+
+  func measureInstructions(_ baselineName: StaticString = #function, block: () -> Void, file: StaticString = #file, line: UInt = #line) throws {
+    let strippedBaselineName = "\(baselineName)".replacingOccurrences(of: "()", with: "")
+    var baseline: UInt64? = try baseline(for: strippedBaselineName)
+
+    if let environmentBaseline = ProcessInfo.processInfo.environment[strippedBaselineName] {
+      baseline = UInt64(environmentBaseline)
+    }
+
+    let startInstructions = getInstructionsExecuted()
+    block()
+    let endInstructions = getInstructionsExecuted()
+    let numberOfInstructions = endInstructions - startInstructions
+
+    guard let baseline = baseline else {
+      XCTFail("Missing baseline for \(strippedBaselineName) with number of instructions \(numberOfInstructions)", file: file, line: line)
+      return
+    }
+
+    let lowerBaseline = UInt64(Double(baseline) * 0.97)
+    let upperBaseline = UInt64(Double(baseline) * 1.03)
```

Comment:
I suspect that we could tighten this a lot further. Maybe to 1% or even 0.5% because the instruction counts are very stable. I recently had a case where I could measure performance differences of <1% https://github.com/apple/swift-syntax/pull/1909/commits/f9982f0959fd4043122a00ebb2a03bb5b674281e

Reply:
Sounds good. Once the PR is done,  I would like to run CI on it 3 times to check that the performance numbers are always within these bounds.